pkg/modifycontroller/controller.go

```diff
@@ -0,0 +1,378 @@
+		name:                    name,
+		modifier:                modifier,
+		kubeClient:              kubeClient,
+		pvSynced:                pvInformer.Informer().HasSynced,
+		pvcSynced:               pvcInformer.Informer().HasSynced,
+		vacSynced:               vacInformer.Informer().HasSynced,
+		claimQueue:              claimQueue,
+		volumes:                 pvInformer.Informer().GetStore(),
+		claims:                  pvcInformer.Informer().GetStore(),
+		volumeAttributesClasses: vacInformer.Informer().GetStore(),
+		eventRecorder:           eventRecorder,
+	}
+
+	// Add a resync period as the PVC's request size can be resized again when we handling
+	// a previous resizing request of the same PVC.
+	pvcInformer.Informer().AddEventHandlerWithResyncPeriod(cache.ResourceEventHandlerFuncs{
+		AddFunc:    ctrl.addPVC,
+		UpdateFunc: ctrl.updatePVC,
+		DeleteFunc: ctrl.deletePVC,
+	}, resyncPeriod)
+
+	return ctrl
+}
+
+func (ctrl *modifyController) initUncertainPVCs() error {
+	ctrl.uncertainPVCs = make(map[string]v1.PersistentVolumeClaim)
+	for _, pvcObj := range ctrl.claims.List() {
+		pvc := pvcObj.(*v1.PersistentVolumeClaim)
+		if pvc.Status.ModifyVolumeStatus != nil && (pvc.Status.ModifyVolumeStatus.Status == v1.PersistentVolumeClaimModifyVolumeInProgress || pvc.Status.ModifyVolumeStatus.Status == v1.PersistentVolumeClaimModifyVolumeInfeasible) {
+			pvcKey, err := cache.MetaNamespaceKeyFunc(pvc)
+			if err != nil {
+				return err
+			}
+			ctrl.uncertainPVCs[pvcKey] = *pvc.DeepCopy()
+		}
+	}
+
+	return nil
+}
+
+func (ctrl *modifyController) addPVC(obj interface{}) {
+	objKey, err := getObjectKey(obj)
+	if err != nil {
+		return
+	}
+	ctrl.claimQueue.Add(objKey)
+}
+
+func (ctrl *modifyController) updatePVC(oldObj, newObj interface{}) {
+	oldPVC, ok := oldObj.(*v1.PersistentVolumeClaim)
+	if !ok || oldPVC == nil {
+		return
+	}
+
+	newPVC, ok := newObj.(*v1.PersistentVolumeClaim)
+	if !ok || newPVC == nil {
+		return
+	}
+
+	// Only trigger modify volume if the following conditions are met
+	// 1. Feature gate is enabled
+	// 2. Non empty vac name
+	// 3. oldVacName != newVacName
+	// 4. PVC is in Bound state
+	if utilfeature.DefaultFeatureGate.Enabled(features.VolumeAttributesClass) {
+		oldVacName := oldPVC.Spec.VolumeAttributesClassName
+		newVacName := newPVC.Spec.VolumeAttributesClassName
+		if newVacName != nil && *newVacName != "" && *newVacName != *oldVacName && oldPVC.Status.Phase == v1.ClaimBound {
+			volumeObj, exists, err := ctrl.volumes.GetByKey(oldPVC.Spec.VolumeName)
+			if err != nil {
+				klog.Errorf("Get PV %q of pvc %q failed: %v", oldPVC.Spec.VolumeName, klog.KObj(oldPVC), err)
+				return
+			}
+			if !exists {
+				klog.InfoS("PV bound to PVC not found", "PV", oldPVC.Spec.VolumeName, "PVC", klog.KObj(oldPVC))
+				return
+			}
+
+			pv, ok := volumeObj.(*v1.PersistentVolume)
+			if !ok {
+				klog.Errorf("expected volume but got %+v", volumeObj)
+				return
+			}
+
+			_, _, err, _ = ctrl.modify(newPVC, pv)
```

Comment:
there's a race. please put it into queue. don't handle it in this func.

Reply:
Do you mean handle it in the sync function?